### PR TITLE
Feature/check no inflight jobs

### PR DESCRIPTION
**Pull request by cookel2**
# What has changed

Changes have been made for the purpose of the following story:

As an API user
I do not want to trigger multiple reindexes in a short space of time
so that the backend processing of a reindex job has time to successfully finish before a new reindex job can be triggered

These are the changes:

The endpoint that allows a POST request to be made, which creates a search reindex job document in mongoDB, has been amended so that the new job will only get created if there are no jobs currently in progress.

If a job is currently in progress then it will contain the following value of state:

"state": "in-progress"

So the endpoint now creates an iteration of all the jobs that have a state set to "in-progress" and orders it so that the one with the most recent reindex_started time is first.

It uses a new environment variable named MAX_REINDEX_JOB_RUNTIME, which gives a duration in seconds, and subtracts this from the current time to calculate a "check from" time.

It then loops through the iteration and checks each reindex_started time to see if it falls between the "check from" time and now. If it does then a 409 conflict error occurs with the following message:

"existing reindex job in progress"

A unit test has been added for the POST request handler.

A few other changes have been made to refactor existing code and increase the coverage of both unit tests and component tests.

# How to review

The endpoint can be tested locally using this POST URL: http://localhost:25700/jobs

The unit tests can be run using this command: make test

The component tests can be run using this command: go test -component

# Who can review

Anyone but me

### CODE REVIEW

File: mongo/job_store.go [hunk 1]
```diff
@@ -38,13 +40,37 @@ func (m *JobStore) CreateJob(ctx context.Context, id string) (job models.Job, er
...
+	result := models.Job{}
+
+	// Check that there are no jobs in progress, which started between a calculated "check from" time and now.
+	// The checkFromTime is calculated by subtracting the configured variable "MaxReindexJobRuntime" from now.
+	checkFromTime := time.Now().Add(-1 * m.cfg.MaxReindexJobRuntime)
+	var jobStartTime time.Time
+	for iter.Next(&result) {
+		//if the start time of the job in progress is later than the checkFromTime but earlier than now then a new job should not be created yet.
+		jobStartTime = result.ReindexStarted
+		if jobStartTime.After(checkFromTime) && jobStartTime.Before(time.Now()) {
```
Comment (ChrysmOre) [inline mongo/job_store.go:59]: Hmm, would it ever be possible for jobStartTime to not be before now?

File: mongo/job_store.go [hunk 2]
```diff
@@ -38,13 +40,37 @@ func (m *JobStore) CreateJob(ctx context.Context, id string) (job models.Job, er
...
+			log.Event(ctx, "found job in progress", log.Data{"id": result.ID, "state": result.State, "start time": jobStartTime})
+			return models.Job{}, ErrExistingJobInProgress
+		}
+	}
+	defer func() {
+		err := iter.Close()
```
Comment (ChrysmOre) [inline mongo/job_store.go:65]: This could be rewritten as one line `if err:= iter.Close; err != ni...`

**Review by ChrysmOre — COMMENTED** [review]
LGTM, couple of small comments

File: mongo/job_store.go (hunk 1 above)
Comment (cookel2) [reply mongo/job_store.go:59]: I think that's a question for my tech lead. I can manually set it to a future time but whether that would actually ever happen I don't know - I'll ask.

File: mongo/job_store.go (hunk 2 above)
Comment (cookel2) [reply mongo/job_store.go]: nice - done

File: mongo/job_store.go (hunk 1 above)
Comment (cookel2) [reply mongo/job_store.go:59]: It looks like it's possible that the jobStartTime could be (at least manually) set to a future time so I'll leave this as is unless you strongly object.

**Review by ChrysmOre — APPROVED** [review]
Tested functionality works as expected
Unit tests pass
Component tests pass

**Pull request merged**